core/src/main/java/com/linecorp/armeria/common/DefaultAttributeMap.java

```diff
@@ -0,0 +1,453 @@
+        final DefaultAttribute<T> result = setAttr(key, value, true);
+        if (result.getValue() != value) {
+            return result.getValue();
+        }
+        return null;
+    }
+
+    <T> void setAttr(AttributeKey<T> key, @Nullable T value) {
+        setAttr(key, value, false);
+    }
+
+    private <T> DefaultAttribute<T> setAttr(AttributeKey<T> key, @Nullable T value, boolean setIfAbsent) {
+        requireNonNull(key, "key");
+        AtomicReferenceArray<DefaultAttribute<?>> attributes = this.attributes;
+        if (attributes == null) {
+            attributes = new AtomicReferenceArray<>(BUCKET_SIZE);
+
+            if (!updater.compareAndSet(this, null, attributes)) {
+                attributes = this.attributes;
+            }
+        }
+        assert attributes != null;
+
+        final int i = index(key);
+        DefaultAttribute<?> head = attributes.get(i);
+        if (head == null) {
+
+            // No head exists yet which means we may be able to add the attribute without synchronization and
+            // just use compare and set. At worst, we need to fallback to synchronization and waste two
+            // allocations.
+            head = new DefaultAttribute<>();
+            final DefaultAttribute<T> attr = new DefaultAttribute<>(key, value);
+            head.next = attr;
+            if (attributes.compareAndSet(i, null, head)) {
+                // We were able to add it so finish the job.
+                return attr;
+            }
+
+            head = attributes.get(i);
+        }
+
+        synchronized (head) {
+            DefaultAttribute<?> curr = head;
+            for (;;) {
+                final DefaultAttribute<?> next = curr.next;
+                if (next != null && next.key == key) {
+                    @SuppressWarnings("unchecked")
+                    final DefaultAttribute<T> attr = (DefaultAttribute<T>) next;
+                    if (setIfAbsent && next.getValue() != null) {
+                        return attr;
+                    }
+
+                    attr.setValue(value);
+                    return attr;
+                }
+
+                if (next == null) {
+                    final DefaultAttribute<T> attr = new DefaultAttribute<>(key, value);
+                    curr.next = attr;
+                    return attr;
+                }
+
+                curr = next;
+            }
+        }
+    }
+
+    Iterator<Entry<AttributeKey<?>, Object>> attrs() {
+        final AtomicReferenceArray<DefaultAttribute<?>> attributes = this.attributes;
+        if (attributes == null) {
+            if (rootAttributeMap == null) {
+                return Collections.emptyIterator();
+            }
+
+            final Iterator<Entry<AttributeKey<?>, Object>> rootAttrs = rootAttributeMap.attrs();
+            if (!rootAttrs.hasNext()) {
+                return Collections.emptyIterator();
+            }
+
+            return new CopyOnWriteAttributeIterator(rootAttrs);
```

Comment:
`CopyOnWriteIterator`? (this class is all about attributes, so..)

Reply:
Fixed. 😉